server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/BaseStarTreeBuilder.java

```diff
@@ -236,18 +254,135 @@ public void build(
             String dimension = dimensionsSplitOrder.get(i).getField();
             FieldInfo dimensionFieldInfo = writeState.fieldInfos.fieldInfo(dimension);
             if (dimensionFieldInfo == null) {
-                dimensionFieldInfo = getFieldInfo(dimension, DocValuesType.SORTED_NUMERIC);
+                // TODO : make this better
+                dimensionFieldInfo = getFieldInfo(dimension, dimensionsSplitOrder.get(i).getDocValuesType());
             }
-            dimensionReaders[i] = new SequentialDocValuesIterator(
-                new SortedNumericStarTreeValuesIterator(fieldProducerMap.get(dimensionFieldInfo.name).getSortedNumeric(dimensionFieldInfo))
+            dimensionReaders[i] = getSequentialDocValuesIterator(
+                dimensionFieldInfo,
+                fieldProducerMap,
+                dimensionsSplitOrder.get(i).getDocValuesType()
             );
+
+            if (dimensionsSplitOrder.get(i).getDocValuesType().equals(DocValuesType.SORTED_SET)) {
+                // This is needed as we need to write the ordinals and also the bytesRef associated with it
+                sortedSetDocValuesMap.put(
+                    dimensionsSplitOrder.get(i).getField(),
+                    fieldProducerMap.get(dimensionFieldInfo.name).getSortedSet(dimensionFieldInfo)
+                );
+            }
+
         }
         Iterator<StarTreeDocument> starTreeDocumentIterator = sortAndAggregateSegmentDocuments(dimensionReaders, metricReaders);
         logger.debug("Sorting and aggregating star-tree in ms : {}", (System.currentTimeMillis() - startTime));
         build(starTreeDocumentIterator, fieldNumberAcrossStarTrees, starTreeDocValuesConsumer);
         logger.debug("Finished Building star-tree in ms : {}", (System.currentTimeMillis() - startTime));
     }
 
+    /**
+     * Copy of BitsFilteredTermsEnum from DocValuesConsumer
+     */
+    static class BitsFilteredTermsEnum extends FilteredTermsEnum {
```

Comment:
should we have a separate class for this? Is Utils package a right place for this?
Another option is if we want to make a wrapper around this like for other lucene classes

Reply:
Its a very small class, will move it to utils.